File: tool/gen_icon.dart
// Run: dart tool/gen_icon.dart
// Generates Ishbor briefcase launcher icons in all Android mipmap densities
// using only dart:ui (no external packages needed).

import 'dart:io';
import 'dart:typed_data';
import 'dart:ui' as ui;

// PNG encoder (hand-written minimal, no package dependency)
Uint8List _encodePng(int width, int height, Uint8List rgba) {
  // We delegate to Flutter's dart:ui Image encoder instead
  throw UnimplementedError('use encodeImage');
}

Future<void> main() async {
  const sizes = {
    'mipmap-mdpi': 48,
    'mipmap-hdpi': 72,
    'mipmap-xhdpi': 96,
    'mipmap-xxhdpi': 144,
    'mipmap-xxxhdpi': 192,
  };

  for (final entry in sizes.entries) {
    final size = entry.value.toDouble();
    final recorder = ui.PictureRecorder();
    final canvas = ui.Canvas(
      recorder,
      ui.Rect.fromLTWH(0, 0, size, size),
    );

    _drawIcon(canvas, size);

    final picture = recorder.endRecording();
    final image = await picture.toImage(entry.value, entry.value);
    final byteData = await image.toByteData(format: ui.ImageByteFormat.png);
    final bytes = byteData!.buffer.asUint8List();

    final path = 'android/app/src/main/res/${entry.key}/ic_launcher.png';
    File(path).writeAsBytesSync(bytes);
    print('Written $path  (${entry.value}x${entry.value})');
  }
  print('Done!');
}

void _drawIcon(ui.Canvas canvas, double s) {
  // Background gradient — deep blue
  final bgPaint = ui.Paint()
    ..shader = ui.Gradient.linear(
      ui.Offset(0, 0),
      ui.Offset(s, s),
      [const ui.Color(0xFF1B4FD8), const ui.Color(0xFF0EA5E9)],
    );

  final bgRRect = ui.RRect.fromRectAndRadius(
    ui.Rect.fromLTWH(0, 0, s, s),
    ui.Radius.circular(s * 0.22),
  );
  canvas.drawRRect(bgRRect, bgPaint);

  final white = ui.Paint()
    ..color = const ui.Color(0xFFFFFFFF)
    ..style = ui.PaintingStyle.fill;

  final whiteSt = ui.Paint()
    ..color = const ui.Color(0xFFFFFFFF)
    ..style = ui.PaintingStyle.stroke
    ..strokeWidth = s * 0.045
    ..strokeCap = ui.StrokeCap.round
    ..strokeJoin = ui.StrokeJoin.round;

  // ── Briefcase / diplomat body ──────────────────────────────────────
  final bx = s * 0.18; // left
  final by = s * 0.42; // top
  final bw = s * 0.64; // width
  final bh = s * 0.36; // height
  final br = s * 0.07; // corner radius

  final bodyPath = ui.Path()
    ..addRRect(
      ui.RRect.fromRectAndRadius(
        ui.Rect.fromLTWH(bx, by, bw, bh),
        ui.Radius.circular(br),
      ),
    );
  canvas.drawPath(
      bodyPath,
      ui.Paint()
        ..color = const ui.Color(0xFFFFFFFF)
        ..style = ui.PaintingStyle.stroke
        ..strokeWidth = s * 0.045
        ..strokeCap = ui.StrokeCap.round
        ..strokeJoin = ui.StrokeJoin.round);

  // Center clasp / handle line
  canvas.drawLine(
    ui.Offset(s * 0.50, by),
    ui.Offset(s * 0.50, by + bh),
    whiteSt,
  );

  // Horizontal middle band
  canvas.drawLine(
    ui.Offset(bx, by + bh * 0.48),
    ui.Offset(bx + bw, by + bh * 0.48),
    whiteSt,
  );

  // ── Handle (top) ──────────────────────────────────────────────────
  final hx = s * 0.36;
  final hw = s * 0.28;
  final hBot = by + s * 0.01;
  final hTop = s * 0.26;
  final hr = s * 0.06;

  final handlePath = ui.Path()
    ..moveTo(hx, hBot)
    ..lineTo(hx, hTop + hr)
    ..arcToPoint(
      ui.Offset(hx + hr, hTop),
      radius: ui.Radius.circular(hr),
      clockwise: false,
    )
    ..lineTo(hx + hw - hr, hTop)
    ..arcToPoint(
      ui.Offset(hx + hw, hTop + hr),
      radius: ui.Radius.circular(hr),
      clockwise: false,
    )
    ..lineTo(hx + hw, hBot);

  canvas.drawPath(handlePath, whiteSt);
}

File: tool/gen_icon.js
// node tool/gen_icon.js
// Generates Ishbor briefcase launcher icons for all Android mipmap densities.
// Requires: npm install sharp (already done)

const sharp = require('sharp');
const fs = require('fs');
const path = require('path');

const sizes = {
  'mipmap-mdpi':    48,
  'mipmap-hdpi':    72,
  'mipmap-xhdpi':   96,
  'mipmap-xxhdpi':  144,
  'mipmap-xxxhdpi': 192,
};

function makeSvg(s) {
  const r = Math.round(s * 0.22); // corner radius
  // proportions (relative to icon size)
  const bx = s * 0.14, by = s * 0.40, bw = s * 0.72, bh = s * 0.40;
  const br = s * 0.07;
  const sw = Math.max(2, s * 0.042); // stroke width
  const hx = s * 0.33, hBot = by + sw / 2, hTop = s * 0.22, hw = s * 0.34;
  const hr = s * 0.07;
  const midY = by + bh * 0.5;
  const midX = bx + bw * 0.5;

  return `<svg xmlns="http://www.w3.org/2000/svg" width="${s}" height="${s}" viewBox="0 0 ${s} ${s}">
  <defs>
    <linearGradient id="bg" x1="0" y1="0" x2="${s}" y2="${s}" gradientUnits="userSpaceOnUse">
      <stop offset="0%" stop-color="#1B4FD8"/>
      <stop offset="100%" stop-color="#0EA5E9"/>
    </linearGradient>
  </defs>

  <!-- Background -->
  <rect width="${s}" height="${s}" rx="${r}" ry="${r}" fill="url(#bg)"/>

  <!-- Briefcase body -->
  <rect x="${bx}" y="${by}" width="${bw}" height="${bh}" rx="${br}" ry="${br}"
        fill="none" stroke="white" stroke-width="${sw}" stroke-linejoin="round"/>

  <!-- Center vertical divider -->
  <line x1="${midX}" y1="${by}" x2="${midX}" y2="${by + bh}"
        stroke="white" stroke-width="${sw}" stroke-linecap="round"/>

  <!-- Horizontal clasp band -->
  <line x1="${bx}" y1="${midY}" x2="${bx + bw}" y2="${midY}"
        stroke="white" stroke-width="${sw}" stroke-linecap="round"/>

  <!-- Handle (arch) -->
  <path d="M ${hx},${hBot}
           L ${hx},${hTop + hr}
           A ${hr},${hr} 0 0,1 ${hx + hr},${hTop}
           L ${hx + hw - hr},${hTop}
           A ${hr},${hr} 0 0,1 ${hx + hw},${hTop + hr}
           L ${hx + hw},${hBot}"
        fill="none" stroke="white" stroke-width="${sw}"
        stroke-linecap="round" stroke-linejoin="round"/>
</svg>`;
}

async function generate() {
  const baseDir = path.join(__dirname, '..', 'android', 'app', 'src', 'main', 'res');
  for (const [folder, size] of Object.entries(sizes)) {
    const svg = Buffer.from(makeSvg(size));
    const outPath = path.join(baseDir, folder, 'ic_launcher.png');
    await sharp(svg)
      .resize(size, size)
      .png()
      .toFile(outPath);
    console.log(`✅  ${outPath}  (${size}x${size})`);
  }

  // Also update web icons
  const webIconDir = path.join(__dirname, '..', 'web', 'icons');
  for (const [webSize, size] of [[192, 192], [512, 512]]) {
    const svg = Buffer.from(makeSvg(size));
    const outPath = path.join(webIconDir, `Icon-${webSize}.png`);
    await sharp(svg).resize(size, size).png().toFile(outPath);
    console.log(`✅  ${outPath}  (${size}x${size})`);
    // Maskable too
    const maskPath = path.join(webIconDir, `Icon-maskable-${webSize}.png`);
    await sharp(svg).resize(size, size).png().toFile(maskPath);
    console.log(`✅  ${maskPath}`);
  }
  // favicon
  const favicon = Buffer.from(makeSvg(64));
  const faviconPath = path.join(__dirname, '..', 'web', 'favicon.png');
  await sharp(favicon).resize(64, 64).png().toFile(faviconPath);
  console.log(`✅  ${faviconPath}`);

  console.log('\nDone! All icons generated.');
}

generate().catch(e => { console.error(e); process.exit(1); });
